### PR TITLE
Allow users to set the trigger height for lengthy toot auto-collapse

### PR DESCRIPTION
**Pull request by neatchee**
Add a field in the glitch-soc preferences to set the exact height in pixels of a "lengthy toot" where auto-collapse is triggered

Originally authored by Dean Bassett (github.com/deanveloper)

Squashed 3 commits from neatchee/mastodon and returned some values to project defaults:
* ef665c1df5821e684c8da3392049f33243fafa74
* 0fce108d210efe55027a3af061bfc57aaaa83843
* 998f701a2b2e37edbda7dffb11a61f67f5559b18

### CODE REVIEW

File: app/javascript/flavours/glitch/components/status.js [hunk 1]
```diff
@@ -251,10 +251,15 @@ class Status extends ImmutablePureComponent {
     // as it could cause surprising changes when receiving notifications
     if (settings.getIn(['content_warnings', 'shared_state']) && status.get('spoiler_text').length && !status.get('hidden')) return;
 
+    let autoCollapseHeight = parseInt(autoCollapseSettings.get('height'));
+    if (status.get('media_attachments').size && !muted) {
+      autoCollapseHeight += 210;
```
Comment (neatchee) [inline app/javascript/flavours/glitch/components/status.js:256]: The actual computed height of the media gallery in advanced mode is 212px.
In simple mode, media-gallery height varies widely.

Default value was 250; dropping this to 210 gives a user experience closer to expectations.

File: app/javascript/flavours/glitch/locales/en.json [hunk 2]
```diff
@@ -103,6 +103,7 @@
   "settings.auto_collapse_all": "Everything",
   "settings.auto_collapse_lengthy": "Lengthy toots",
   "settings.auto_collapse_media": "Toots with media",
+  "settings.auto_collapse_height": "Height (pixels) of lengthy toots",
```
Comment (Plastikmensch) [inline app/javascript/flavours/glitch/locales/en.json:106]: I think this is unclear.
I actually prefer the wording in the earlier PRs: "Maximum height (pixels) of toots before auto-collapsing",
or "Height (in pixels) to consider a toot lengthy" (or similar).

**Review by Plastikmensch — CHANGES_REQUESTED** [review]
I still think using pixels isn't the user friendliest option, but I'm also not opposed to features for more tech savvy users.

File: app/javascript/flavours/glitch/features/local_settings/page/index.js [hunk 3]
```diff
@@ -406,6 +406,18 @@ class LocalSettingsPage extends React.PureComponent {
...
           </LocalSettingsPageItem>
+	  <LocalSettingsPageItem
+            settings={settings}
+            item={['collapsed', 'auto', 'height']}
+            id='mastodon-settings--collapsed-auto-height'
+            placeholder='400'
+	    onChange={onChange}
+            dependsOn={[['collapsed', 'enabled']]}
+            dependsOnNot={[['collapsed', 'auto', 'all']]}
+            inputProps={{ type: 'number', min: '200', max: '999', width: '45px'}}
```
[the remark this answers is not in the document]
Comment (neatchee) [reply app/javascript/flavours/glitch/features/local_settings/page/index.js:417]: No, this definitely works. I've tested it locally.

[the remark this answers is not in the document]
Comment (Plastikmensch) [reply app/javascript/flavours/glitch/features/local_settings/page/index.js:417]: Ah, I see what's happening
Firefox:
Screenshot_2023-01-06_19-08-07

Chromium:
Screenshot_2023-01-06_19-09-50

[the remark this answers is not in the document]
Comment (neatchee) [reply app/javascript/flavours/glitch/features/local_settings/page/index.js:417]: Hooray for browser inconsistency. I'll move this to CSS

File: app/javascript/flavours/glitch/locales/en.json (hunk 2 above)
Comment (neatchee) [reply app/javascript/flavours/glitch/locales/en.json:106]: I like the latter suggestion. The previous language was unclear as to it's relationship to the "lengthy toots" option so your suggestion works great

**Comment by neatchee** [comment]
> I still think using pixels isn't the user friendliest option, but I'm also not opposed to features for more tech savvy users.

Changing this to something other than pixels is a wider-scoped change, since that's how the feature itself is designed. I'm generally disinclined to do static math behind the scenes (any style font change would throw this off), and making it actually respond to line-height feels overly complicated.

I don't really have strong feelings either way. If others do, we can change it :)

File: app/javascript/flavours/glitch/components/status.js (hunk 1 above)
Comment (Plastikmensch) [reply app/javascript/flavours/glitch/components/status.js:256]: I'm not opposed to this change, but it's also browser specific and even timeline specific.
For example in chrome it is 212px in home, but 209px in the local timeline.
In Firefox it is 219px in home and 216px in the local timeline.

Comment (neatchee) [reply app/javascript/flavours/glitch/components/status.js:256]: Either way it sounds like 250 isn't the right value. To my mind, taking the lower end of the range makes sense; I've enabled this setting to clean up my timeline, presumably. Being on the aggressive side of collapsing makes sense for the user intent.

**Review by Plastikmensch — APPROVED** [review]

File: app/javascript/flavours/glitch/components/status.js (hunk 1 above)
Comment (Plastikmensch) [reply app/javascript/flavours/glitch/components/status.js:256]: I agree.
Would be good to change the comment above the change to reflect the change.

```js  
//   -  The user has decided to collapse long statuses and the status is
//      over 400px (without media, or 650px with).
```

to

```js
  //   -  The user has decided to collapse long statuses and the status is
  //      over the user set value (400px by default without media, or 610px with).
```

Other than that I don't see anything that would prevent merging this.

File: app/javascript/flavours/glitch/features/local_settings/page/index.js [hunk 4]
```diff
@@ -406,6 +406,18 @@ class LocalSettingsPage extends React.PureComponent {
           >
             <FormattedMessage id='settings.auto_collapse_media' defaultMessage='Toots with media' />
           </LocalSettingsPageItem>
+	  <LocalSettingsPageItem
+            settings={settings}
+            item={['collapsed', 'auto', 'height']}
+            id='mastodon-settings--collapsed-auto-height'
+            placeholder='400'
+	    onChange={onChange}
```
Comment (ClearlyClaire) [inline app/javascript/flavours/glitch/features/local_settings/page/index.js:414]: Why the inconsistent indentation?

File: app/javascript/flavours/glitch/styles/components/local_settings.scss [hunk 5]
```diff
@@ -110,6 +110,10 @@
       text-decoration: none;
     }
   }
+
+  #mastodon-settings--collapsed-auto-height {
+    width: 45px
+  }
```
Comment (ClearlyClaire) [inline app/javascript/flavours/glitch/styles/components/local_settings.scss:116]: Why a specific style for this?

Comment (neatchee) [reply app/javascript/flavours/glitch/styles/components/local_settings.scss:116]: This field looked ridiculous without having its width restricted, and InputProps does not support the width parameter for all browsers

It's there another way you'd prefer it done?

File: app/javascript/flavours/glitch/features/local_settings/page/index.js (hunk 4 above)
Comment (neatchee) [reply app/javascript/flavours/glitch/features/local_settings/page/index.js:414]: Unintended. Resolving.

File: app/javascript/flavours/glitch/styles/components/local_settings.scss (hunk 5 above)
Comment (ClearlyClaire) [reply app/javascript/flavours/glitch/styles/components/local_settings.scss:116]: Oh, I see! A hardcoded pixel width doesn't sound safe though wrt. potentially different fonts and browser layout. I would suggest to use `em` or `ch` but that is not quite right either as the size needs to include the up/down arrows, which is browser-dependent… `calc(3ch + 20px)` is equivalent to your proposal but should scale a bit depending on the used font… but it's no guarantee either. Also we may want to allow space for 4 characters, so `calc(4ch + 20px)`.

**Pull request merged**